### PR TITLE
Conditional assert

### PR DESCRIPTION
**Pull request by pmconne**
Ensure assertions have as little overhead as possible unless they have been enabled, either by setting NODE_ENV=development or explicitly enabling them via `setAssertionsEnabled`.
Previously, we relied on our webpack tools to strip calls to `assert` from the transpiled JavaScript, but that never worked properly, stopped working entirely recently, and only could work if apps use our build tools/configuration.
See revised documentation of `assert`.

### CODE REVIEW

**Comment by kabentley** [comment]
what about renaming this to something other than `assert`? Node has `assert`, console has `assert`, chai has `assert`, there's asserts in TypesScript?

**Comment by pmconne** [comment]
> what about renaming this to something other than `assert`?

That sounds like an entirely separate PR.

File: core/bentley/src/Assert.ts [hunk 1]
```diff
@@ -17,7 +19,55 @@
  * which indicate a misuse of the API which should be eliminated during development.
  * @public
  */
-export function assert(condition: boolean, msg?: string): asserts condition {
-  if (!condition)
-    throw new Error(`Assert: ${msg ?? "Programmer Error"}`);
+
+/** Asserts that a condition is `true` and - in development builds - throws an error if it is not.
```
Comment (kabentley) [inline core/bentley/src/Assert.ts:23]: "in development builds" -> "when enabled"

File: core/bentley/src/Assert.ts [hunk 2]
```diff
@@ -17,7 +19,55 @@
...
+ * By default, `assert` does nothing unless the environment variable `NODE_ENV` exists and is set to `development`. Use [[setAssertionsEnabled]] to override this behavior.
+ *
+ * If your condition and/or message is relatively expensive to compute (e.g., more complex than just a plain boolean or string), pass them as functions to prevent them
+ * from being evaluated when assertions are disabled.
+ *
+ * @param condition The condition that is asserted to be `true`.
+ * @param message An optional description of the condition being asserted, to be included in the exception if `condition` is `false`. Defaults to "Programmer Error".
+ * @throws Error containing the specified `message` if `condition` is `false`.
+ * @see [[setAssertionsEnabled]] to enable or disable assertions.
+ * @public
```
Comment (kabentley) [inline core/bentley/src/Assert.ts:43]: I think a note that, as opposed to assert in C++, "this assert is not entirely free even when disabled" is warranted.

File: core/bentley/src/Assert.ts [hunk 3]
```diff
@@ -17,7 +19,55 @@
...
+ * Assertions should **never** be used to test for conditions - however unlikely - that could be expected to occur at run-time,
+ * such as failing to write to a file or load a resource over the network. If the condition asserted ever fails in a production environment,
+ * the programmer has made a serious mistake.
+ *
+ * By default, `assert` does nothing unless the environment variable `NODE_ENV` exists and is set to `development`. Use [[setAssertionsEnabled]] to override this behavior.
+ *
+ * If your condition and/or message is relatively expensive to compute (e.g., more complex than just a plain boolean or string), pass them as functions to prevent them
+ * from being evaluated when assertions are disabled.
+ *
+ * @param condition The condition that is asserted to be `true`.
```
Comment (kabentley) [inline core/bentley/src/Assert.ts:39]: or a function that evaluates to the condition.  The advice paragraph above about when to use a function should be part of this @param doc.

**Review by kabentley — APPROVED** [review]

File: core/bentley/src/Assert.ts (hunk 1 above)
Comment (pmconne) [reply core/bentley/src/Assert.ts]: fixed.

File: core/bentley/src/Assert.ts (hunk 2 above)
Comment (pmconne) [reply core/bentley/src/Assert.ts:31]: Clarified.

File: core/bentley/src/Assert.ts (hunk 3 above)
Comment (pmconne) [reply core/bentley/src/Assert.ts]: Done.

**Review by bbastings — APPROVED** [review]

**Review by mgooding — APPROVED** [review]

**Review by aruniverse — APPROVED** [review]
fyi @markschlosseratbentley

File: core/bentley/src/Assert.ts [hunk 4]
```diff
@@ -6,18 +6,56 @@
...
+ * When [[assert]] is disabled, it returns immediately without evaluating the asserted condition. By default, it is enabled only if
+ * the environment variable `NODE_ENV` exists and is set to `development`.
+ * @param enabled If true, `assert` will throw upon a false assertion.
+ * @returns true if assertions were enabled prior to this call.
  * @public
  */
-export function assert(condition: boolean, msg?: string): asserts condition {
-  if (!condition)
-    throw new Error(`Assert: ${msg ?? "Programmer Error"}`);
+export function setAssertionsEnabled(enabled: boolean): boolean {
```
Comment (wgoehrig) [inline core/bentley/src/Assert.ts:57]: Is this really necessary?  I just did some testing, and without defining this function, Terser is able to do some really impressive optimizations - basically it completely remove calls to assert without any special configuration.

But as soon as it sees this function definition, it (rightly) determines that it can no longer safely assume all that code can be removed.

**Review by wgoehrig — CHANGES_REQUESTED** [review]

File: core/bentley/src/Assert.ts (hunk 4 above)
Comment (wgoehrig) [reply core/bentley/src/Assert.ts:57]: Here's an example:

```ts
import { setAssertionsEnabled, assert } from "./Assert";
assert(someTest());
```
would currently webpack/minify down to (with whitespace re-added):
```js
(window.webpackJsonp = window.webpackJsonp || []).push([
    [0],
    [function(o, e, r) {
        o.exports = r(1)
    }, function(o, e, r) {
        "use strict";
        r.r(e);
        let n = !1;
        ! function(o, e) {
            if (n && ("boolean" != typeof o && (o = o()), !o)) throw "string" != typeof(e = e || "Programmer Error") && (e = e()), new Error("Assert: {message}")
        }(someTest())
    }],
    [
        [0, 1]
    ]
]);
```
but as soon as I delete the definition for `setAssertionsEnabled`, that same example bundled becomes:
```js
(window.webpackJsonp = window.webpackJsonp || []).push([
    [0],
    [function(n, o, s) {
        n.exports = s(1)
    }, function(n, o, s) {
        "use strict";
        s.r(o);
        someTest()
    }],
    [
        [0, 1]
    ]
]);
```

Comment (aruniverse) [reply core/bentley/src/Assert.ts:57]: Paul, is there a scenario where you'd want to turn off asserts in development?

Comment (wgoehrig) [reply core/bentley/src/Assert.ts:57]: Yeah, @pmconne, just to clarify a bit, for frontend builds, `process.env.NODE_ENV` is usually replaced with a compile-time constant (either `"development"` or `"production"`).  This is a truly bonkers convention, but a fairly well-established one, and most bundling tools support it out of the box.

Comment (pmconne) [reply core/bentley/src/Assert.ts:57]: You're going to have to tell me what I'm looking at there, @wgoehrig.
It looks to me like the call to `assert` is being elided at bundle time, which is not what I want. I want assertions enabled or disabled at run-time based on NODE_ENV.

Comment (pmconne) [reply core/bentley/src/Assert.ts:57]: @wgoehrig that means that apps who consume are packages will never see assertions from our packages' code. That is not entirely desirable.

Comment (wgoehrig) [reply core/bentley/src/Assert.ts:57]: That's the crucial intermediate step here tho - `process.env.NODE_ENV` becomes `"production"`, so

```
let assertionsEnabled = process.env.NODE_ENV === "development";
```
effectively becomes
```
let assertionsEnabled = false;
```
and Terser takes over from there.  Personally, I was surprised that just removing `setAssertionsEnabled` worked - I would have assumed you'd at least also needed to make `assertionsEnabled` const, but I guess Terser was able to see that `assertionsEnabled` wasn't exported and nothing in the file was left to ever mutate it...

Comment (wgoehrig) [reply core/bentley/src/Assert.ts:57]: Sorry, GH is not refreshing these fast enough and I keep responding to older comments.

@pmconne the convention everyone uses is basically the following:

- `npm start` does a hot-reloading _development_ build, so asserts are still present
- `npm run build` does a production build, which minifies, disables asserts, etc.

Since all this minification only happens when a whole _app_ is bundled, consumers of our packages will still see asserts during development of their app.

Comment (pmconne) [reply core/bentley/src/Assert.ts:57]: So @wgoehrig, talk to me like a webpack idiot, because I am: if somebody installs core-frontend from npm and bundles it with their app, they will see assertions from core-frontend as well as from their own code IFF process.env.NODE_ENV is defined when they bundle their app?

Comment (pmconne) [reply core/bentley/src/Assert.ts:57]: @wgoehrig yeah, we crossed streams again.
Given that, the only purpose of `setAssertionsEnabled` is tests. The tests were useful for testing these changes but I'm ok with removing them now.

Comment (pmconne) [reply core/bentley/src/Assert.ts:57]: Unless we can set NODE_ENV=development when we webpack the tests?

Comment (wgoehrig) [reply core/bentley/src/Assert.ts:57]: Almost.  The value of _your_ process.env.NODE_ENV at build time is not actually relevant - the constant injected into the code comes purely from the [webpack configuration](https://webpack.js.org/configuration/optimization/#optimizationnodeenv).  Again, using process.env in the name there is super misleading and downright bizarre, but that's the established convention, and many of _our_ dependencies require it.

So basically everyone just has two different webpack configs, and they use them for the different npm scripts.

Comment (wgoehrig) [reply core/bentley/src/Assert.ts:57]: > Unless we can set NODE_ENV=development when we webpack the tests?

Yeah, that should be easy enough - I'll look into that.

Comment (pmconne) [reply core/bentley/src/Assert.ts:57]: @wgoehrig nah don't worry about it-  most of the tests want to turn assertions on and off anyway.
Just to be clear: what was process.NODE_ENV set to when you produced the terser output?

Comment (wgoehrig) [reply core/bentley/src/Assert.ts:57]: Actually, good news - all our test webpack configs are already setting that.

Comment (wgoehrig) [reply core/bentley/src/Assert.ts:57]: I did not have `process.NODE_ENV` set at all when I produced either of those outputs, but my webpack config had [mode](https://webpack.js.org/configuration/mode/) set to "production".

If I explicitly override the config to set NODE_ENV to "development" (but still optimize/minify - not a usual combination), I get:

```js
(window.webpackJsonp = window.webpackJsonp || []).push([
    [0],
    [function(o, r, e) {
        o.exports = e(1)
    }, function(o, r, e) {
        "use strict";
        e.r(r);
        ! function(o, r) {
            if ("boolean" != typeof o && (o = o()), !o) throw "string" != typeof(r = r || "Programmer Error") && (r = r()), new Error("Assert: {message}")
        }(someTest())
    }],
    [
        [0, 1]
    ]
]);
"
```

Comment (pmconne) [reply core/bentley/src/Assert.ts:57]: @wgoehrig then help me understand. Is `someTest()` being invoked below?
```
(window.webpackJsonp = window.webpackJsonp || []).push([
    [0],
    [function(n, o, s) {
        n.exports = s(1)
    }, function(n, o, s) {
        "use strict";
        s.r(o);
        someTest()
    }],
    [
        [0, 1]
    ]
]);
```

**Comment by pmconne** [comment]
@wgoehrig are you certain `process.NODE_ENV` is set to "development" when we build the bentley-core tests?
Updated  tests:
```
describe("assert", () => {
  it("only throws if condition is false", () => {
    expect(() => assert(false)).to.throw("Assert: Programmer Error");
    expect(() => assert(true)).not.to.throw("Assert: Programmer Error");
  });

  it("includes message in error if supplied", () => {
    expect(() => assert(false, "ruh-roh!")).to.throw("Assert: ruh-roh!");
  });

  it("accepts a function for condition or message", () => {
    expect(() => assert(() => false)).to.throw("Assert: Programmer Error");
    expect(() => assert(() => true)).not.to.throw;
    expect(() => assert(false, () => "Danger Will Robinson!")).to.throw("Assert: Danger Will Robinson!");
  });

  it("only evaluates message function if condition is false", () => {
    let called = false;
    const message = () => { called = true; return "message"; };

    assert(true, message);
    expect(called).to.be.false;

    try {
      assert(false, message);
    } catch (_) { }
    expect(called).to.be.true;
  });
});
```

Results:
```
  1) assert
       only throws if condition is false:
     AssertionError: expected [Function] to throw an error
      at Context.<anonymous> (src\test\Assert.test.ts:10:41)
      at processImmediate (internal/timers.js:464:21)

  2) assert
       includes message in error if supplied:
     AssertionError: expected [Function] to throw an error
      at Context.<anonymous> (src\test\Assert.test.ts:15:53)
      at processImmediate (internal/timers.js:464:21)

  3) assert
       accepts a function for condition or message:
     AssertionError: expected [Function] to throw an error
      at Context.<anonymous> (src\test\Assert.test.ts:19:47)
      at processImmediate (internal/timers.js:464:21)

  4) assert
       only evaluates message function if condition is false:

      AssertionError: expected false to be true
      + expected - actual

      -false
      +true

      at Context.<anonymous> (src\test\Assert.test.ts:34:25)
      at processImmediate (internal/timers.js:464:21)
```

**Comment by pmconne** [comment]
```
describe("assert", () => {
  it("is enabled", () => {
    expect(process.env.NODE_ENV).to.equal("development");
  });
});
```

```
  1) assert
       is enabled:
     AssertionError: expected undefined to equal 'development'
      at Context.<anonymous> (src\test\Assert.test.ts:10:37)
      at processImmediate (internal/timers.js:464:21)
```

File: core/bentley/src/Assert.ts (hunk 4 above)
Comment (wgoehrig) [reply core/bentley/src/Assert.ts:57]: Short answer is yes.  Terser didn't know whether `someTest()` had side effects and so that wasn't safe to remove, even when eliding the call to `assert`.

All the examples I gave are incomplete because there's actually another file with the rest of the webpack boilerplate.  Webpack now defaults to splitting that into a separate file because it can be cached client-side longer - it never needs to change when the rest of your code changes.  It's also total gibberish once minified, but basically it finds the "entry" module function in webpackJsonp and calls it.

So it's the _other_ file that will eventually call that function above (with `"use strict";` in the body).

Comment (pmconne) [reply core/bentley/src/Assert.ts]: Wonderful. I've removed `setAssertionsEnabled`, removed the tests, and updated the docs.

**Comment by wgoehrig** [comment]
@pmconne sorry - I meant to say that all _certa_ tests than need to be bundled w/ webpack are setting that.  Something like core-bentley tests **will** actually use the run-time process.env.NODE_ENV

**Comment by wgoehrig** [comment]
So for those tests I think the easiest thing to do would be to set NODE_ENV in the package.json scripts.

Or maybe we just want to change it to be:

```ts
let assertionsEnabled = process.env.NODE_ENV !== "production";
```

File: core/bentley/src/Assert.ts [hunk 5]
```diff
@@ -6,9 +6,10 @@
  * @module Utils
  */
 
-let assertionsEnabled = process.env.NODE_ENV === "development";
+const assertionsEnabled = process.env.NODE_ENV === "development";
 
-/** Asserts that a condition is `true` and - when enabled - throws an error if it is not.
+/** Asserts that a condition is `true` and - when enabled - throws an error if it is not. Assertions are enabled only if `process.NODE_ENV` is set to `development`.
```
Comment (wgoehrig) [inline core/bentley/src/Assert.ts]:
```suggestion
/** Asserts that a condition is `true` and - when enabled - throws an error if it is not. Assertions are enabled only if `process.env.NODE_ENV` is set to `development`.
```

Comment (pmconne) [reply core/bentley/src/Assert.ts]: fixed.

**Comment by wgoehrig** [comment]
> So for those tests I think the easiest thing to do would be to set NODE_ENV in the package.json scripts.
> 
> Or maybe we just want to change it to be:
> 
> ```ts
> let assertionsEnabled = process.env.NODE_ENV !== "production";
> ```

@pmconne - want me to go thru and set NODE_ENV in package.json test scripts, or would you rather just invert `assertionsEnabled`?

**Comment by pmconne** [comment]
> @pmconne - want me to go thru and set NODE_ENV in package.json test scripts, or would you rather just invert `assertionsEnabled`?

I deleted the tests for `assert` so I think changing package.json scripts is unnecessary. I don't think we should invert `assertionsEnabled`.

**Review by wgoehrig — APPROVED** [review]
